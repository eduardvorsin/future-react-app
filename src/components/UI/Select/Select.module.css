.select {
	display: flex;
	align-items: center;
	justify-content: space-between;
	flex-wrap: wrap;
}

.select__label {
	color: var(--light-color-200);
	margin-bottom: 10px;
	margin-right: 0;
	flex: 0 1 100%;
}

.select__wrapper {
	flex: 0 1 100%;
}

.select-wrapper {
	position: relative;
	border: 2px solid transparent;
	width: 100%;
	height: 38px;
	border-radius: 4px;
	font-size: 18px;
	cursor: pointer;
	line-height: 1.1;
	background-color: var(--light-color-200);
	transition: border var(--transition-time) var(--transition-easing);

}

.select-wrapper::after {
	position: absolute;
	top: 50%;
	right: 6px;
	transform: translateY(-50%);
	content: "";
	width: 13px;
	height: 6px;
	background-color: var(--brand-color-200);
	clip-path: polygon(100% 0%, 0 0%, 50% 100%);
}

.select-wrapper:hover:not([disabled]) {
	border: 2px solid var(--brand-color-200);
	transition: border var(--transition-time) var(--transition-easing);
}

.select-wrapper__select {
	appearance: none;
	background-color: transparent;
	border: none;
	padding: 4px 24px 4px 8px;
	margin: 0;
	width: 100%;
	font-family: inherit;
	font-size: inherit;
	cursor: inherit;
	line-height: inherit;
	height: 100%;
}

@media(min-width:576px) {
	.select {
		flex-wrap: nowrap;
	}

	.select__label {
		flex: 0 1 30%;
		margin-bottom: 0;
		margin-right: 10px;
	}

	.select__wrapper {
		height: 48px;
		flex: 0 1 70%;
	}
}