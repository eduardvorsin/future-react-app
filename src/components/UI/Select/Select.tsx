import React, { SelectHTMLAttributes, memo } from 'react';
import classes from './Select.module.css';

type SelectOption = {
  value: string,
  label: string,
}
type SelectBaseProps = Partial<SelectHTMLAttributes<HTMLSelectElement>>;
interface SelectProps extends SelectBaseProps {
  className?: string,
  labelText: string,
  id: string,
  options: SelectOption[],
  value: string
  onChange: React.ChangeEventHandler<HTMLSelectElement>,
}

const Select: React.FC<SelectProps> = ({
  className,
  labelText,
  id,
  options,
  value,
  onChange,
}) => {
  const selectClasses = [
    className,
    classes.select,
  ].join(' ');

  const selectChangeHandler: React.ChangeEventHandler<HTMLSelectElement> = (e) => {
    onChange(e);
  };

  return (
    <div
      className={selectClasses}
    >
      <label
        className={classes.select__label}
        htmlFor={id}
      >
        {labelText}
      </label>
      <div
        className={classes['select-wrapper']}
      >
        <select
          className={classes['select-wrapper__select']}
          value={value}
          onChange={selectChangeHandler}
          id={id}
        >
          {options.map((option) => (
            <option
              key={option.value}
              value={option.value}
            >
              {option.label}
            </option>
          ))}
        </select>
      </div>
    </div>
  );
};

export const MemoSelect = memo(Select);

export default Select;
