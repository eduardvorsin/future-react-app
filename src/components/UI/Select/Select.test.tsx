import React, { ChangeEvent } from 'react';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import Select from './Select';

const mockData = [
  {
    value: 'test-value-1',
    label: 'test-label-1',
  },
  {
    value: 'test-value-2',
    label: 'test-label-2',
  },
  {
    value: 'test-value-3',
    label: 'test-label-3',
  },
  {
    value: 'test-value-4',
    label: 'test-label-4',
  },
  {
    value: 'test-value-5',
    label: 'test-label-5',
  },
];

describe('Select tests', () => {
  it('is rendered correctly', () => {
    const mockFn = jest.fn<void, [e: ChangeEvent]>();

    render(
      <Select
        id='test-select'
        labelText='test-select'
        options={mockData}
        onChange={mockFn}
        value={mockData[0].value}
        testId='test-select'
      />,
    );

    expect(screen.getByTestId<HTMLDivElement>('test-select')).toBeInTheDocument();
    expect(screen.getByRole<HTMLSelectElement>('combobox')).toBeInTheDocument();
    expect(screen.getAllByRole<HTMLOptionElement>('option')).toHaveLength(5);
  });
  it('the mock function should be triggered when the selection item is changed', async () => {
    const user = userEvent.setup();
    const mockFn = jest.fn<void, [e: ChangeEvent]>();

    render(
      <Select
        id='test-select'
        labelText='test-select'
        options={mockData}
        onChange={mockFn}
        value={mockData[0].value}
        testId='test-select'
      />,
    );

    await user.selectOptions(screen.getByRole<HTMLSelectElement>('combobox'), 'test-value-2');

    expect(mockFn).toHaveBeenCalledTimes(1);
  });

  it('if the options array is empty then the select is rendered without options', () => {
    const mockFn = jest.fn<void, [e: ChangeEvent]>();

    render(
      <Select
        id='test-select'
        labelText='test-select'
        options={[]}
        onChange={mockFn}
        value={mockData[0].value}
        testId='test-select'
      />,
    );

    expect(screen.queryByRole<HTMLOptionElement>('option')).not.toBeInTheDocument();
  });

  it('is a basic snapshot', () => {
    const mockFn = jest.fn<void, [e: ChangeEvent]>();

    render(
      <Select
        id='test-select'
        labelText='test-select'
        options={mockData}
        onChange={mockFn}
        value={mockData[0].value}
        testId='test-select'
      />,
    );

    expect(screen.getByTestId<HTMLDivElement>('test-select')).toMatchSnapshot();
  });

  it('is asnapshot with an empty array of options', () => {
    const mockFn = jest.fn<void, [e: ChangeEvent]>();

    render(
      <Select
        id='test-select'
        labelText='test-select'
        options={[]}
        onChange={mockFn}
        value={mockData[0].value}
        testId='test-select'
      />,
    );

    expect(screen.getByTestId<HTMLDivElement>('test-select')).toMatchSnapshot();
  });
});
