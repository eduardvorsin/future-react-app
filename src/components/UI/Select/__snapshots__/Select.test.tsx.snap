// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select tests is a basic snapshot 1`] = `
<div
  class=" select"
  data-testid="test-select"
>
  <label
    class="select__label"
    for="test-select"
  >
    test-select
  </label>
  <div
    class="select-wrapper"
  >
    <select
      class="select-wrapper__select"
      id="test-select"
    >
      <option
        value="test-value-1"
      >
        test-label-1
      </option>
      <option
        value="test-value-2"
      >
        test-label-2
      </option>
      <option
        value="test-value-3"
      >
        test-label-3
      </option>
      <option
        value="test-value-4"
      >
        test-label-4
      </option>
      <option
        value="test-value-5"
      >
        test-label-5
      </option>
    </select>
  </div>
</div>
`;

exports[`Select tests is asnapshot with an empty array of options 1`] = `
<div
  class=" select"
  data-testid="test-select"
>
  <label
    class="select__label"
    for="test-select"
  >
    test-select
  </label>
  <div
    class="select-wrapper"
  >
    <select
      class="select-wrapper__select"
      id="test-select"
    />
  </div>
</div>
`;
